backend/blog/views.py

```diff
@@ -21,7 +20,9 @@ def blog_post(request, slug):
     """
     Single blog post view
     """
-    post = BlogPost.objects.get(slug=slug)
-    tags = Tag.objects.all()
-    context = {'post': post, 'tags': tags}
+    posts = BlogPost.objects.filter(slug=slug)
+    if not posts or (not posts[0].published and request.user != posts[0].author):
+        return render(request, 'blog/blog_404.html')  # Respond with 404 page
+    tags = posts[0].tags.names()
+    context = {'post': posts[0], 'tags': tags}
     return render(request, 'blog/post.html', context)
```

Comment:
In post.html, let's add a banner at the top if the post is unpublished.
Something like:
```THIS IS A PREVIEW ONLY. TO SEE THIS LIVE, CLICK "published" IN THE ADMIN PANEL```